File: core/test/TestFixtures/Trading.hs
module TestFixtures.Trading where
import TestFixtures.GameState
import Test.Hspec

import Domain.GameState
import Domain.Resource
import Domain.Player
import Core.Trading

import Data.Map(fromList)

testTrading = do
    describe "Trading.tryTrading" $ do
        context "nothing to trade" $ do
            let trade = tryTrading gameStateForTradingEmpty "b"
            it "1" $ trade []                                                         `shouldBe` []
            it "2" $ trade [[Cost Wood 1]]                                            `shouldBe` []
            it "3" $ trade [[Cost Wood 2]]                                            `shouldBe` []
            it "4" $ trade [[Cost Wood 1, Cost Loom 1]]                               `shouldBe` []
            it "5" $ trade [[Cost Wood 1], [Cost Loom 1]]                             `shouldBe` []
            it "6" $ trade [[Cost Wood 1, Cost Loom 2], [Cost Ore 3, Cost Papyrus 4]] `shouldBe` []
        context "trade without discount" $ do
            let trade = tryTrading gameStateForTrading0 "b"
            it "1" $ trade [[Cost Wood 1]]               `shouldBe` [[("c", Wood, 2)]]
            it "2" $ trade [[Cost Ore 1]]                `shouldBe` [[("a", Ore, 2)]]
            it "3" $ trade [[Cost Clay 1]]               `shouldBe` [[("a", Clay, 2)], [("c", Clay, 2)]]
            it "4" $ trade [[Cost Stone 1]]              `shouldBe` []
            it "5" $ trade [[Cost Wood 1], [Cost Ore 1]] `shouldBe` [[("c", Wood, 2)], [("a", Ore, 2)]]
            it "6" $ trade [[Cost Wood 2]]               `shouldBe` []
            it "7" $ trade [[Cost Ore 2]]                `shouldBe` []
            it "8" $ trade [[Cost Clay 2]]               `shouldBe` [[("a", Clay, 2), ("c", Clay, 2)]]
            it "9" $ trade [[Cost Wood 1, Cost Ore 1]]   `shouldBe` [[("a", Ore, 2), ("c", Wood, 2)]]
            it "10" $ trade [[Cost Clay 3]]              `shouldBe` []
        context "trade with one discount on all resources" $ do
            let trade = tryTrading gameStateForTrading1 "b"
            it "1" $ trade [[Cost Wood 1]]               `shouldBe` [[("c", Wood, 2)]]
            it "2" $ trade [[Cost Ore 1]]                `shouldBe` [[("a", Ore, 1)]]
            it "3" $ trade [[Cost Clay 1]]               `shouldBe` [[("a", Clay, 1)], [("c", Clay, 2)]]
            it "4" $ trade [[Cost Stone 1]]              `shouldBe` []
            it "5" $ trade [[Cost Wood 1], [Cost Ore 1]] `shouldBe` [[("c", Wood, 2)], [("a", Ore, 1)]]
            it "6" $ trade [[Cost Wood 2]]               `shouldBe` []
            it "7" $ trade [[Cost Ore 2]]                `shouldBe` []
            it "8" $ trade [[Cost Clay 2]]               `shouldBe` [[("a", Clay, 1), ("c", Clay, 2)]]
            it "9" $ trade [[Cost Wood 1, Cost Ore 1]]   `shouldBe` [[("a", Ore, 1), ("c", Wood, 2)]]
            it "10" $ trade [[Cost Clay 3]]              `shouldBe` []
        context "trade with both discounts on all resources" $ do
            let trade = tryTrading gameStateForTrading2 "b"
            it "1" $ trade [[Cost Wood 1]]               `shouldBe` [[("c", Wood, 1)]]
            it "2" $ trade [[Cost Ore 1]]                `shouldBe` [[("a", Ore, 1)]]
            it "3" $ trade [[Cost Clay 1]]               `shouldBe` [[("a", Clay, 1)], [("c", Clay, 1)]]
            it "4" $ trade [[Cost Stone 1]]              `shouldBe` []
            it "5" $ trade [[Cost Wood 1], [Cost Ore 1]] `shouldBe` [[("c", Wood, 1)], [("a", Ore, 1)]]
            it "6" $ trade [[Cost Wood 2]]               `shouldBe` []
            it "7" $ trade [[Cost Ore 2]]                `shouldBe` []
            it "8" $ trade [[Cost Clay 2]]               `shouldBe` [[("a", Clay, 1), ("c", Clay, 1)]]
            it "9" $ trade [[Cost Wood 1, Cost Ore 1]]   `shouldBe` [[("a", Ore, 1), ("c", Wood, 1)]]
            it "10" $ trade [[Cost Clay 3]]              `shouldBe` []
        context "trade with one discount on some resources" $ do
            let trade = tryTrading gameStateForTrading3 "b"
            it "1" $ trade [[Cost Wood 1]]               `shouldBe` [[("c", Wood, 2)]]
            it "2" $ trade [[Cost Ore 1]]                `shouldBe` [[("a", Ore, 2)]]
            it "3" $ trade [[Cost Stone 1]]              `shouldBe` []
            it "4" $ trade [[Cost Clay 1]]               `shouldBe` [[("a", Clay, 1)], [("c", Clay, 2)]]
            it "5" $ trade [[Cost Wood 1], [Cost Ore 1]] `shouldBe` [[("c", Wood, 2)], [("a", Ore, 2)]]
            it "6" $ trade [[Cost Wood 2]]               `shouldBe` []
            it "7" $ trade [[Cost Ore 2]]                `shouldBe` []
            it "8" $ trade [[Cost Clay 2]]               `shouldBe` [[("a", Clay, 1), ("c", Clay, 2)]]
            it "9" $ trade [[Cost Wood 1, Cost Ore 1]]   `shouldBe` [[("a", Ore, 2), ("c", Wood, 2)]]
            it "10" $ trade [[Cost Clay 3]]              `shouldBe` []
        context "trade without discount repeating resources" $ do
            let trade = tryTrading gameStateForTrading4 "b"
            it "1" $ trade [[Cost Clay 1]] `shouldBe` [[("a", Clay, 2)], [("c", Clay, 2)]]
            it "2" $ trade [[Cost Clay 2]] `shouldBe` [[("a", Clay, 2), ("c", Clay, 2)], [("c", Clay, 2), ("c", Clay, 2)]]
            it "3" $ trade [[Cost Clay 3]] `shouldBe` [[("a", Clay, 2), ("c", Clay, 2), ("c", Clay, 2)]]
    describe "Trading.diffResourceCosts" $ do
        it "1" $ []                        `diffResourceCosts` []            `shouldBe` []
        it "2" $ [Cost Wood 1]             `diffResourceCosts` [Cost Wood 1] `shouldBe` []
        it "3" $ [Cost Wood 1]             `diffResourceCosts` [Cost Loom 1] `shouldBe` [Cost Wood 1]
        it "4" $ [Cost Wood 1, Cost Ore 1] `diffResourceCosts` [Cost Wood 1] `shouldBe` [Cost Ore 1]
        it "5" $ [Cost Wood 1, Cost Ore 1] `diffResourceCosts` [Cost Ore 1]  `shouldBe` [Cost Wood 1]
        it "6" $ [Cost Wood 3]             `diffResourceCosts` [Cost Wood 2] `shouldBe` [Cost Wood 1]
        it "6" $ [Cost Wood 3]             `diffResourceCosts` [Cost Wood 4] `shouldBe` []

gameStateForTradingEmpty =
    defaultGameState {
        neighbours = initNeighbours ["a", "b", "c"],
        playerStates = fromList [
            ("a", (defaultPlayerState "a") {
                resourcesProduced = []
            }),
            ("b", (defaultPlayerState "b") {
                tradeActions = [defaultTradeAction]
            }),
            ("c", (defaultPlayerState "c") {
                resourcesProduced = []
            })
        ]
    }

gameStateForTrading0 =
    defaultGameState {
        neighbours = initNeighbours ["a", "b", "c"],
        playerStates = fromList [
            ("a", (defaultPlayerState "a") {
                resourcesProduced = [Single Ore 1, Single Clay 1]
            }),
            ("b", (defaultPlayerState "b") {
                tradeActions = [defaultTradeAction]
            }),
            ("c", (defaultPlayerState "c") {
                resourcesProduced = [Single Clay 1, Single Wood 1]
            })
        ]
    }

gameStateForTrading1 =
    defaultGameState {
        neighbours = initNeighbours ["a", "b", "c"],
        playerStates = fromList [
            ("a", (defaultPlayerState "a") {
                resourcesProduced = [Single Ore 1, Single Clay 1]
            }),
            ("b", (defaultPlayerState "b") {
                tradeActions = [
                    defaultTradeAction,
                    return tradeAction :: Action (Domain.Player.Name -> ResourceType -> Int)
                ]
            }),
            ("c", (defaultPlayerState "c") {
                resourcesProduced = [Single Clay 1, Single Wood 1]
            })
        ]
    }
    where
        tradeAction :: Domain.Player.Name -> ResourceType -> Int
        tradeAction "a" = const 1
        tradeAction _  = const 2

gameStateForTrading2 =
    defaultGameState {
        neighbours = initNeighbours ["a", "b", "c"],
        playerStates = fromList [
            ("a", (defaultPlayerState "a") {
                resourcesProduced = [Single Ore 1, Single Clay 1]
            }),
            ("b", (defaultPlayerState "b") {
                tradeActions = [
                    defaultTradeAction,
                    return $ const $ const 1 :: Action (Domain.Player.Name -> ResourceType -> Int)
                ]
            }),
            ("c", (defaultPlayerState "c") {
                resourcesProduced = [Single Clay 1, Single Wood 1]
            })
        ]
    }

gameStateForTrading3 =
    defaultGameState {
        neighbours = initNeighbours ["a", "b", "c"],
        playerStates = fromList [
            ("a", (defaultPlayerState "a") {
                resourcesProduced = [Single Ore 1, Single Clay 1]
            }),
            ("b", (defaultPlayerState "b") {
                tradeActions = [
                    defaultTradeAction,
                    return tradeAction :: Action (Domain.Player.Name -> ResourceType -> Int)
                ]
            }),
            ("c", (defaultPlayerState "c") {
                resourcesProduced = [Single Clay 1, Single Wood 1]
            })
        ]
    }
    where
        tradeAction :: Domain.Player.Name -> ResourceType -> Int
        tradeAction "a" Clay = 1
        tradeAction _ _ = 2

gameStateForTrading4 =
    defaultGameState {
        neighbours = initNeighbours ["a", "b", "c"],
        playerStates = fromList [
            ("a", (defaultPlayerState "a") {
                resourcesProduced = [Single Ore 1, Single Clay 1]
            }),
            ("b", (defaultPlayerState "b") {
                tradeActions = [defaultTradeAction]
            }),
            ("c", (defaultPlayerState "c") {
                resourcesProduced = [Single Clay 2]
            })
        ]
    }

File: core/src/tests/trading.rs
use super::helpers::{default_game_state, default_player_state};
use crate::core::trading::{try_trading, TradingOptions};
use crate::domain::{
    GameState, Player, PlayerState, ResourceCost, ResourceType::*, ResourcesProduced,
};
use maplit::{hashmap, hashset};

#[test]
fn test_nothing_to_trade() {
    let f = |resource_cost_options: &'static [&'static [ResourceCost]]| -> TradingOptions {
        let game_state = GameState {
            player_states: hashmap! {
                "a".to_string() => PlayerState {
                    player: Player("a".to_string()),
                    ..default_player_state()
                },
                "b".to_string() => PlayerState {
                    player: Player("b".to_string()),
                    ..default_player_state()
                }.add_trade_action_move(Box::new(|_, _| 2)),
                "c".to_string() => PlayerState {
                    player: Player("c".to_string()),
                    ..default_player_state()
                },
            },
            ..default_game_state()
        };
        try_trading(
            &game_state,
            &"b".to_string(),
            resource_cost_options.iter().map(|r| r.to_vec()).collect(),
        )
    };

    assert_eq!(f(&[]), hashset![]);
    assert_eq!(f(&[&[ResourceCost(Wood, 1)]]), hashset![]);
    assert_eq!(f(&[&[ResourceCost(Wood, 2)]]), hashset![]);
    assert_eq!(
        f(&[&[ResourceCost(Wood, 1), ResourceCost(Loom, 1)]]),
        hashset![]
    );
    assert_eq!(
        f(&[&[ResourceCost(Wood, 1)], &[ResourceCost(Loom, 1)]]),
        hashset![]
    );
    assert_eq!(
        f(&[
            &[ResourceCost(Wood, 1), ResourceCost(Loom, 2)],
            &[ResourceCost(Ore, 3), ResourceCost(Papyrus, 4)]
        ]),
        hashset![]
    );
}

#[test]
fn test_trade_without_discount() {
    let f = |resource_cost_options: &'static [&'static [ResourceCost]]| -> TradingOptions {
        let game_state = GameState {
            player_states: hashmap! {
                "a".to_string() => PlayerState {
                    player: Player("a".to_string()),
                    resources_produced: ResourcesProduced {
                        single_resources: hashmap! {
                            Ore => 1,
                            Clay => 1
                        },
                        ..Default::default()
                    },
                    ..default_player_state()
                },
                "b".to_string() => PlayerState {
                    player: Player("b".to_string()),
                    ..default_player_state()
                }.add_trade_action_move(Box::new(|_, _| 2)),
                "c".to_string() => PlayerState {
                    player: Player("c".to_string()),
                    resources_produced: ResourcesProduced {
                        single_resources: hashmap! {
                            Wood => 1,
                            Clay => 1
                        },
                        ..Default::default()
                    },
                    ..default_player_state()
                },
            },
            ..default_game_state()
        };
        try_trading(
            &game_state,
            &"b".to_string(),
            resource_cost_options.iter().map(|r| r.to_vec()).collect(),
        )
    };

    assert_eq!(
        f(&[&[ResourceCost(Wood, 1)]]),
        hashset![vec![("c".to_string(), Wood, 2)]]
    );
    assert_eq!(
        f(&[&[ResourceCost(Ore, 1)]]),
        hashset![vec![("a".to_string(), Ore, 2)]]
    );
    assert_eq!(
        f(&[&[ResourceCost(Clay, 1)]]),
        hashset![
            vec![("a".to_string(), Clay, 2)],
            vec![("c".to_string(), Clay, 2)]
        ]
    );
    assert_eq!(f(&[&[ResourceCost(Stone, 1)]]), hashset![]);
    assert_eq!(
        f(&[&[ResourceCost(Wood, 1)], &[ResourceCost(Ore, 1)]]),
        hashset![
            vec![("c".to_string(), Wood, 2)],
            vec![("a".to_string(), Ore, 2)]
        ]
    );
    assert_eq!(f(&[&[ResourceCost(Wood, 2)]]), hashset![]);
    assert_eq!(f(&[&[ResourceCost(Ore, 2)]]), hashset![]);
    assert_eq!(
        f(&[&[ResourceCost(Clay, 2)]]),
        hashset![vec![("a".to_string(), Clay, 2), ("c".to_string(), Clay, 2)]]
    );
    assert_eq!(
        f(&[&[ResourceCost(Wood, 1), ResourceCost(Ore, 1)]]),
        hashset![vec![("a".to_string(), Ore, 2), ("c".to_string(), Wood, 2)]]
    );
    assert_eq!(f(&[&[ResourceCost(Clay, 3)]]), hashset![]);
}

#[test]
fn test_trade_with_one_discount_on_all_resources() {
    let f = |resource_cost_options: &'static [&'static [ResourceCost]]| -> TradingOptions {
        let game_state = GameState {
            player_states: hashmap! {
                "a".to_string() => PlayerState {
                    player: Player("a".to_string()),
                    resources_produced: ResourcesProduced {
                        single_resources: hashmap! {
                            Ore => 1,
                            Clay => 1
                        },
                        ..Default::default()
                    },
                    ..default_player_state()
                },
                "b".to_string() => PlayerState {
                    player: Player("b".to_string()),
                    ..default_player_state()
                }.add_trade_action_move(Box::new(|player_name, _| {
                    if player_name == "a" {
                        1
                    } else {
                        2
                    }
                })),
                "c".to_string() => PlayerState {
                    player: Player("c".to_string()),
                    resources_produced: ResourcesProduced {
                        single_resources: hashmap! {
                            Wood => 1,
                            Clay => 1
                        },
                        ..Default::default()
                    },
                    ..default_player_state()
                },
            },
            ..default_game_state()
        };
        try_trading(
            &game_state,
            &"b".to_string(),
            resource_cost_options.iter().map(|r| r.to_vec()).collect(),
        )
    };

    assert_eq!(
        f(&[&[ResourceCost(Wood, 1)]]),
        hashset![vec![("c".to_string(), Wood, 2)]]
    );
    assert_eq!(
        f(&[&[ResourceCost(Ore, 1)]]),
        hashset![vec![("a".to_string(), Ore, 1)]]
    );
    assert_eq!(
        f(&[&[ResourceCost(Clay, 1)]]),
        hashset![
            vec![("a".to_string(), Clay, 1)],
            vec![("c".to_string(), Clay, 2)]
        ]
    );
    assert_eq!(f(&[&[ResourceCost(Stone, 1)]]), hashset![]);
    assert_eq!(
        f(&[&[ResourceCost(Wood, 1)], &[ResourceCost(Ore, 1)]]),
        hashset![
            vec![("c".to_string(), Wood, 2)],
            vec![("a".to_string(), Ore, 1)]
        ]
    );
    assert_eq!(f(&[&[ResourceCost(Wood, 2)]]), hashset![]);
    assert_eq!(f(&[&[ResourceCost(Ore, 2)]]), hashset![]);
    assert_eq!(
        f(&[&[ResourceCost(Clay, 2)]]),
        hashset![vec![("a".to_string(), Clay, 1), ("c".to_string(), Clay, 2)]]
    );
    assert_eq!(
        f(&[&[ResourceCost(Wood, 1), ResourceCost(Ore, 1)]]),
        hashset![vec![("a".to_string(), Ore, 1), ("c".to_string(), Wood, 2)]]
    );
    assert_eq!(f(&[&[ResourceCost(Clay, 3)]]), hashset![]);
}

#[test]
fn test_trade_with_both_discounts_on_all_resources() {
    let f = |resource_cost_options: &'static [&'static [ResourceCost]]| -> TradingOptions {
        let game_state = GameState {
            player_states: hashmap! {
                "a".to_string() => PlayerState {
                    player: Player("a".to_string()),
                    resources_produced: ResourcesProduced {
                        single_resources: hashmap! {
                            Ore => 1,
                            Clay => 1
                        },
                        ..Default::default()
                    },
                    ..default_player_state()
                },
                "b".to_string() => PlayerState {
                    player: Player("b".to_string()),
                    ..default_player_state()
                }.add_trade_action_move(Box::new(|_, _| 1)),
                "c".to_string() => PlayerState {
                    player: Player("c".to_string()),
                    resources_produced: ResourcesProduced {
                        single_resources: hashmap! {
                            Wood => 1,
                            Clay => 1
                        },
                        ..Default::default()
                    },
                    ..default_player_state()
                },
            },
            ..default_game_state()
        };
        try_trading(
            &game_state,
            &"b".to_string(),
            resource_cost_options.iter().map(|r| r.to_vec()).collect(),
        )
    };

    assert_eq!(
        f(&[&[ResourceCost(Wood, 1)]]),
        hashset![vec![("c".to_string(), Wood, 1)]]
    );
    assert_eq!(
        f(&[&[ResourceCost(Ore, 1)]]),
        hashset![vec![("a".to_string(), Ore, 1)]]
    );
    assert_eq!(
        f(&[&[ResourceCost(Clay, 1)]]),
        hashset![
            vec![("a".to_string(), Clay, 1)],
            vec![("c".to_string(), Clay, 1)]
        ]
    );
    assert_eq!(f(&[&[ResourceCost(Stone, 1)]]), hashset![]);
    assert_eq!(
        f(&[&[ResourceCost(Wood, 1)], &[ResourceCost(Ore, 1)]]),
        hashset![
            vec![("c".to_string(), Wood, 1)],
            vec![("a".to_string(), Ore, 1)]
        ]
    );
    assert_eq!(f(&[&[ResourceCost(Wood, 2)]]), hashset![]);
    assert_eq!(f(&[&[ResourceCost(Ore, 2)]]), hashset![]);
    assert_eq!(
        f(&[&[ResourceCost(Clay, 2)]]),
        hashset![vec![("a".to_string(), Clay, 1), ("c".to_string(), Clay, 1)]]
    );
    assert_eq!(
        f(&[&[ResourceCost(Wood, 1), ResourceCost(Ore, 1)]]),
        hashset![vec![("a".to_string(), Ore, 1), ("c".to_string(), Wood, 1)]]
    );
    assert_eq!(f(&[&[ResourceCost(Clay, 3)]]), hashset![]);
}

#[test]
fn test_trade_with_one_discount_on_some_resources() {
    let f = |resource_cost_options: &'static [&'static [ResourceCost]]| -> TradingOptions {
        let game_state = GameState {
            player_states: hashmap! {
                "a".to_string() => PlayerState {
                    player: Player("a".to_string()),
                    resources_produced: ResourcesProduced {
                        single_resources: hashmap! {
                            Ore => 1,
                            Clay => 1
                        },
                        ..Default::default()
                    },
                    ..default_player_state()
                },
                "b".to_string() => PlayerState {
                    player: Player("b".to_string()),
                    ..default_player_state()
                }.add_trade_action_move(Box::new(|player_name, resource_type| {
                    if player_name == "a" && *resource_type == Clay {
                        1
                    } else {
                        2
                    }
                })),
                "c".to_string() => PlayerState {
                    player: Player("c".to_string()),
                    resources_produced: ResourcesProduced {
                        single_resources: hashmap! {
                            Wood => 1,
                            Clay => 1
                        },
                        ..Default::default()
                    },
                    ..default_player_state()
                },
            },
            ..default_game_state()
        };
        try_trading(
            &game_state,
            &"b".to_string(),
            resource_cost_options.iter().map(|r| r.to_vec()).collect(),
        )
    };

    assert_eq!(
        f(&[&[ResourceCost(Wood, 1)]]),
        hashset![vec![("c".to_string(), Wood, 2)]]
    );
    assert_eq!(
        f(&[&[ResourceCost(Ore, 1)]]),
        hashset![vec![("a".to_string(), Ore, 2)]]
    );
    assert_eq!(f(&[&[ResourceCost(Stone, 1)]]), hashset![]);
    assert_eq!(
        f(&[&[ResourceCost(Clay, 1)]]),
        hashset![
            vec![("a".to_string(), Clay, 1)],
            vec![("c".to_string(), Clay, 2)]
        ]
    );
    assert_eq!(
        f(&[&[ResourceCost(Wood, 1)], &[ResourceCost(Ore, 1)]]),
        hashset![
            vec![("c".to_string(), Wood, 2)],
            vec![("a".to_string(), Ore, 2)]
        ]
    );
    assert_eq!(f(&[&[ResourceCost(Wood, 2)]]), hashset![]);
    assert_eq!(f(&[&[ResourceCost(Ore, 2)]]), hashset![]);
    assert_eq!(
        f(&[&[ResourceCost(Clay, 2)]]),
        hashset![vec![("a".to_string(), Clay, 1), ("c".to_string(), Clay, 2)]]
    );
    assert_eq!(
        f(&[&[ResourceCost(Wood, 1), ResourceCost(Ore, 1)]]),
        hashset![vec![("a".to_string(), Ore, 2), ("c".to_string(), Wood, 2)]]
    );
    assert_eq!(f(&[&[ResourceCost(Clay, 3)]]), hashset![]);
}

#[test]
fn test_trade_without_discount_repeating_resources() {
    let f = |resource_cost_options: &'static [&'static [ResourceCost]]| -> TradingOptions {
        let game_state = GameState {
            player_states: hashmap! {
                "a".to_string() => PlayerState {
                    player: Player("a".to_string()),
                    resources_produced: ResourcesProduced {
                        single_resources: hashmap! {
                            Ore => 1,
                            Clay => 1
                        },
                        ..Default::default()
                    },
                    ..default_player_state()
                },
                "b".to_string() => PlayerState {
                    player: Player("b".to_string()),
                    ..default_player_state()
                }.add_trade_action_move(Box::new(|_, _| 2)),
                "c".to_string() => PlayerState {
                    player: Player("c".to_string()),
                    resources_produced: ResourcesProduced {
                        single_resources: hashmap! {
                            Clay => 2
                        },
                        ..Default::default()
                    },
                    ..default_player_state()
                },
            },
            ..default_game_state()
        };
        try_trading(
            &game_state,
            &"b".to_string(),
            resource_cost_options.iter().map(|r| r.to_vec()).collect(),
        )
    };

    assert_eq!(
        f(&[&[ResourceCost(Clay, 1)]]),
        hashset![
            vec![("a".to_string(), Clay, 2)],
            vec![("c".to_string(), Clay, 2)]
        ]
    );
    assert_eq!(
        f(&[&[ResourceCost(Clay, 2)]]),
        hashset![
            vec![("a".to_string(), Clay, 2), ("c".to_string(), Clay, 2)],
            vec![("c".to_string(), Clay, 2), ("c".to_string(), Clay, 2)]
        ]
    );
    assert_eq!(
        f(&[&[ResourceCost(Clay, 3)]]),
        hashset![vec![
            ("a".to_string(), Clay, 2),
            ("c".to_string(), Clay, 2),
            ("c".to_string(), Clay, 2)
        ]]
    );
}
